lxd/storage_zfs.go

```diff
@@ -547,7 +547,73 @@ func (s *storageZfs) StoragePoolUpdate(writable *api.StoragePoolPut, changedConf
 	}
 
 	if shared.StringInSlice("volume.size", changedConfig) {
-		return fmt.Errorf("the \"volume.size\" property cannot be changed")
+		volumes, err := db.StoragePoolVolumesGet(s.d.db, s.poolID,
+			[]int{storagePoolVolumeTypeContainer})
+		if err != nil {
+			if err != db.NoSuchObjectError {
+				return err
+			}
+		}
+
+		// The name of a container is always identical to the name of
+		// its storage volume.
+		for _, volume := range volumes {
+			ct, err := containerLoadByName(s.d, volume.Name)
+			if err != nil {
+				return err
+			}
+
+			// Quota do not apply to snapshots (I *think*.)
+			if ct.IsSnapshot() {
+				continue
+			}
+
+			// Check if the storage volume itself already has local
+			// "size" property set.
+			if volume.Config["size"] != "" {
+				continue
+			}
+
+			// Check if the container has a "size" property set on
+			// its root disk device in its local devices.
+			localDevices := ct.LocalDevices()
+			_, rootDiskDevice, _ := containerGetRootDiskDevice(localDevices)
+			// Check if the container inherits a "size" property
+			// from a root disk device in one of its profiles.
+			if rootDiskDevice["size"] == "" {
+				expandedDevices := ct.ExpandedDevices()
+				_, rootDiskDevice, _ = containerGetRootDiskDevice(expandedDevices)
+			}
+
+			// Non-storage pool "size" properties override storage
+			// pool "size" properties
+			if rootDiskDevice["size"] != "" {
+				continue
+			}
+
+			ourStart, err := ct.StorageStartSensitive()
+			if err != nil {
+				return err
+			}
+			if ourStart {
+				defer ct.StorageStop()
+			}
+
+			size := int64(0)
+			if writable.Config["volume.size"] != "" {
+				size, err = shared.ParseByteSizeString(writable.Config["volume.size"])
+				if err != nil {
+					return err
+				}
+			}
+
+			st := ct.Storage()
+			err = st.StorageEntitySetQuota(storagePoolVolumeTypeContainer,
```

Comment:
So that takes care of updating things when volume.size is changed on the pool.
Do we have code which also update things when the volume itself is directly modified?